File: problems/Swift/384.打乱数组.swift
/**
 * 384.打乱数组 [Medium]
 * Time Complexity: O(n), Space Complexity: O(n)
 * https://leetcode-cn.com/problems/shuffle-an-array/
 */

class Solution {
  private var numbers: [Int]
  private var originals: [Int]

  init(_ nums: [Int]) {
    numbers = nums
    originals = nums
  }

  func reset() -> [Int] {
    numbers = originals
    return numbers
  }

  func shuffle() -> [Int] {
    for i in 0 ..< numbers.count {
      let j = Int.random(in: i ..< numbers.count)
      numbers.swapAt(i, j)
    }
    return numbers
  }
}

/**
 * Your Solution object will be instantiated and called as such:
 * let obj = Solution(nums)
 * let ret_1: [Int] = obj.reset()
 * let ret_2: [Int] = obj.shuffle()
 */

File: problems/rust-solutions/src/384.打乱数组.rs
// 384.打乱数组 [Medium]
// Time Complexity: O(n), Space Complexity: O(n)
// https://leetcode-cn.com/problems/shuffle-an-array/

use rand::prelude::SliceRandom;
use rand::thread_rng;

struct Solution {
  nums: Vec<i32>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Solution {

  fn new(nums: Vec<i32>) -> Self {
    Solution { nums }
  }

  fn reset(&self) -> Vec<i32> {
    return self.nums.clone();
  }

  fn shuffle(&self) -> Vec<i32> {
    let mut ans = self.nums.clone();
    ans.shuffle(&mut thread_rng());
    return ans;
  }

}

/**
 * Your Solution object will be instantiated and called as such:
 * let obj = Solution::new(nums);
 * let ret_1: Vec<i32> = obj.reset();
 * let ret_2: Vec<i32> = obj.shuffle();
 */
